src/coreclr/jit/morph.cpp

```diff
@@ -13780,10 +13780,80 @@ void Compiler::fgMorphBlock(BasicBlock* block)
 
     if (optLocalAssertionProp)
     {
-        // For now, each block starts with an empty table, and no available assertions
-        //
-        optAssertionReset(0);
-        apLocal = BitVecOps::MakeEmpty(apTraits);
+        if (!optCrossBlockLocalAssertionProp)
+        {
+            // Each block starts with an empty table, and no available assertions
+            //
+            optAssertionReset(0);
+            apLocal = BitVecOps::MakeEmpty(apTraits);
+        }
+        else
+        {
+            // Determine if this block can leverage assertions from its pred blocks.
+            //
+            // Some blocks are ineligible.
+            //
+            bool canUsePredAssertions = ((block->bbFlags & BBF_CAN_ADD_PRED) == 0) && !bbIsHandlerBeg(block);
+
+            // Validate all preds have valid info
+            //
+            if (canUsePredAssertions)
+            {
+                bool hasPred = false;
+
+                for (BasicBlock* const pred : block->PredBlocks())
+                {
+                    // A higher postorder number means the block appears earlier in
+                    // the postorder. Use this to detect if a pred's assertion info is available.
+                    //
+                    if (pred->bbPostorderNum > block->bbPostorderNum)
+                    {
+                        // Yes, available. If this is the first pred, copy.
+                        // If this is a subsequent pred, intersect.
+                        //
+                        if (!hasPred)
+                        {
+                            apLocal = BitVecOps::MakeCopy(apTraits, pred->bbAssertionOut);
+                            hasPred = true;
+                        }
+                        else
+                        {
+                            BitVecOps::IntersectionD(apTraits, apLocal, pred->bbAssertionOut);
+                        }
+
+                        continue;
+                    }
+
+                    // No, not available.
+                    //
+                    JITDUMP(FMT_BB " pred " FMT_BB " not processed; clearing assertions in\n", block->bbNum,
+                            pred->bbNum);
+                    canUsePredAssertions = false;
```

Comment:
add `break;`?

Or better, make the failure case first, with `break`, and eliminate the `continue`